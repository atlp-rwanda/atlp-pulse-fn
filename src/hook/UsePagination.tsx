import { useState } from 'react';

const usePagination = (initialSize = 3) => {
  const [pagination, setPagination] = useState({
    pageSize: initialSize,
    pageIndex: 0,
  });

  const { pageSize, pageIndex } = pagination;

  return {
    onPaginationChange: setPagination,
    pagination,
    limit: pageSize,
    skip: pageSize * pageIndex,
  };
};

export default usePagination;
