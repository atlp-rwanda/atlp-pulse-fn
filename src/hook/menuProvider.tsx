/* eslint-disable react/jsx-no-constructed-context-values */
import React, {
  createContext,
  ReactNode,
  useMemo,
  useState,
} from 'react';

interface Props {
  children?: ReactNode;
}

export const MenuContext = createContext<any>({
  minimized: false,
  showNav: true,
});

/* istanbul ignore next */
export function ThemeProvider({ children }: Props) {
  const [minimized, setMinimized] = useState(false);
  const [showNav, setShowNav] = useState(false);

  const toggleMinMenu = () => setMinimized((p) => !p);

  const toggleNav = () => setShowNav((p) => !p);

  const contextValue = useMemo(
    () => ({
      minimized,
      setMinimized,
      toggleMinMenu,
      showNav,
      setShowNav,
      toggleNav,
    }),
    [minimized, setMinimized, toggleMinMenu, showNav, setShowNav, toggleNav],
  );
  return (
    <MenuContext.Provider value={contextValue}>{children}</MenuContext.Provider>
  );
}

export default ThemeProvider;
