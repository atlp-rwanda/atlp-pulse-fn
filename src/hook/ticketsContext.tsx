/* eslint-disable */
/*istanbul ignore file*/
import React, { createContext, ReactNode, useState } from 'react';

interface Props {
  children?: ReactNode;
  // any props that come into the component
}

const initialState: Array<any> | null = null;

export const TicketsContext = createContext<any>(initialState);
/* istanbul ignore next */
export function TicketsProvider({ children }: Props) {
  const [tickets, setTickets] = useState<any>(initialState);
  /* istanbul ignore next */
  const setAllTickets = (tickets: Array<any>) => {
     /* istanbul ignore next */
    setTickets(tickets);
  };
  /* istanbul ignore next */
 /* istanbul ignore next */
  const updateTicketStatus = (ticketId: string, status: string) => {
    const newTickets = tickets.map((ticket: any) => {
      if (ticket.id === ticketId) return { ...ticket, status: status };
      return ticket;
    });
     /* istanbul ignore next */
    setTickets(newTickets);
  };
  /* istanbul ignore next */
  const addReply = (ticketId: string, reply: any, status: string) => {
    const newTickets = tickets.map((ticket: any) => {
      if (ticket.id === ticketId) {
        return { ...ticket, status, replies: [reply, ...ticket.replies] };
      }
      return ticket;
    });
    setTickets(newTickets);
  };
  /* istanbul ignore next */
  const addTicket = (ticket: any) => {
     /* istanbul ignore next */
    setTickets((tickets: Array<any>) => [ticket, ...tickets]);
  };

  return (
    <TicketsContext.Provider
      value={{
        tickets,
        setAllTickets,
        updateTicketStatus,
        addReply,
        addTicket,
      }}
    >
      {children}
    </TicketsContext.Provider>
  );
}

export default TicketsProvider;
