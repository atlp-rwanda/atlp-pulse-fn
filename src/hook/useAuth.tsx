import React, { createContext, ReactNode, useState } from 'react';

const getInitialState = () => {
  if (window && window.localStorage) {
    const storedUser = window.localStorage.getItem('auth');
    if (storedUser) return JSON.parse(storedUser);
  }
  return { name: '', role: 'user', auth: false };
};
interface Props {
  children?: ReactNode;
  // any props that come into the component
}
export const UserContext = createContext<any>(getInitialState);

const UserProvider = ({ children, ...props }: Props) => {
  const [user, setUser] = useState<any>(getInitialState);
  const login = (data: any) => {
    localStorage.setItem(
      'auth',
      JSON.stringify({ name: data.name, auth: true, role: data.role }),
    );
    setUser(() => ({ name: data.name, auth: true, role: data.role }));
  };
  const logout = () => {
    setUser(() => ({ name: '', role: 'user', auth: false }));
  };

  return (
    <UserContext.Provider {...props} value={{ user, login, logout }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
