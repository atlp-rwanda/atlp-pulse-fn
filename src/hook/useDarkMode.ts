import { useState } from 'react';

export type ThemeType = 'dark' | 'light';

export const getInitialTheme = (): ThemeType | any => {
  if (typeof window !== 'undefined' && window.localStorage) {
    const storedPrefs = window.localStorage.getItem('color-theme');
    if (storedPrefs) return storedPrefs;
    const userMedia = window.matchMedia('(prefers-color-scheme: dark)');
    if (userMedia.matches) return 'dark';
    return 'light';
  }
  return 'light';
};

export const setThemeHeader = (newTheme: ThemeType, oldTheme?: ThemeType) => {
  const root = window.document.documentElement;
  localStorage.setItem('color-theme', newTheme);
  if (oldTheme) root.classList.remove(oldTheme);
  root.classList.add(newTheme);
  root.setAttribute('data-theme', newTheme);
};

const useDarkMode = () => {
  const [colorTheme, changeTheme] = useState<ThemeType>(getInitialTheme());
  setThemeHeader(colorTheme);

  return [colorTheme, changeTheme] as const;
};

export default useDarkMode;
