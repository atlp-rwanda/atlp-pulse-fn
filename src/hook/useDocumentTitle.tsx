import { useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import getLanguage from '../utils/getLanguage';

function useDocumentTitle(title: string) {
  const language = getLanguage();
  const { t } = useTranslation();
  useEffect(() => {
    document.title = `${t(title)} | Devpulse`;
  }, [title, language]);
}

export default useDocumentTitle;
