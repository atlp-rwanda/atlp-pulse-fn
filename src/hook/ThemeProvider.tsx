/* eslint-disable react/jsx-no-constructed-context-values */
import React, { createContext, ReactNode, useState } from 'react';
import useDarkMode, { getInitialTheme, setThemeHeader, ThemeType } from './useDarkMode';

interface Props {
  children?: ReactNode;
}

export const ThemeContext = createContext<any>(getInitialTheme);

/* istanbul ignore next */
export function ThemeProvider({ children }: Props) {
  const [colorTheme, changeTheme] = useDarkMode();

  const setTheme = (val: ThemeType) => {
    changeTheme((prev) => {
      setThemeHeader(val, prev);
      return val;
    });
  };

  return (
    <ThemeContext.Provider value={{ colorTheme, setTheme }}>
      {children}
    </ThemeContext.Provider>
  );
}

export default ThemeProvider;
