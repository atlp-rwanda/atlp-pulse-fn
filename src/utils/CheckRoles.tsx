/* eslint-disable no-restricted-syntax */
import React, { ReactNode, useContext } from 'react';
import { UserContext } from '../hook/useAuth';

interface Props {
  children: ReactNode;
  roles: string[];
  // any props that come into the component
}

function filterRoles(arr: string[]) {
  const good: string[] = [];
  const bad: string[] = [];

  for (const el of arr) {
    const value = el.trim();
    if (value[0] === '-') {
      const [_one, ...badRole] = value;
      bad.push(badRole.join('').trim());
    } else {
      good.push(value);
    }
  }

  return { good, bad };
}

function CheckRole({ children, roles, ...props }: Props) {
  const { good, bad } = filterRoles(roles);
  const { user } = useContext(UserContext);
  if (
    roles.length === 0 ||
    (bad.length && !bad.includes(user?.role)) ||
    (good.length && good?.includes(user?.role))
  )
    return <React.Fragment {...props}>{children}</React.Fragment>;
  return <React.Fragment {...props} />;
}

export default CheckRole;
