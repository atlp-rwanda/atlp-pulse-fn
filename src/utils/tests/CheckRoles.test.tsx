/* eslint-disable */
import * as React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { UserContext } from '../../hook/useAuth';

import CheckRoles from '../CheckRoles';

describe('CheckRoles component', () => {
  it('should not render the child component', () => {
    const user = { name: '', role: 'user', auth: false };
    const elem = renderer
      .create(
        <MemoryRouter>
          <UserContext.Provider value={user}>
            <CheckRoles roles={['admin']}>
              <div></div>
            </CheckRoles>
          </UserContext.Provider>
        </MemoryRouter>,
      )
      .toJSON();
    expect(elem).toMatchSnapshot();
  });
  it('should render the the child compoenent', () => {
    const user = { role: 'admin', name: 'Fabrice', auth: true };
    const elem = renderer
      .create(
        <MemoryRouter>
          <UserContext.Provider value={{ user }}>
            <CheckRoles roles={['admin']}>
              <div></div>
            </CheckRoles>
          </UserContext.Provider>
        </MemoryRouter>,
      )
      .toJSON();
    expect(elem).toMatchSnapshot();
  });
});
