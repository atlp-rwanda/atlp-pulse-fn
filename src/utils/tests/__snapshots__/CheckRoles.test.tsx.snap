// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckRoles component should not render the child component 1`] = `null`;

exports[`CheckRoles component should render the the child compoenent 1`] = `<div />`;
